circleci-artifacts/src/main/java/com/palantir/gradle/utils/circleciartifacts/CircleCiArtifacts.java

```
@@ -47,13 +53,47 @@ public final Provider<ArtifactLocation> resolveArtifactLocation(String pathWithi
+                .map(CircleCiArtifacts::extractDomain)
+                .orElse("https://<circle_url>");
+        Provider<String> workflowJobId = getVariables().envVarOrFromTestingProperty("CIRCLE_WORKFLOW_JOB_ID");
 
-        Provider<String> externalLocation = circleProjectUsername
-                .zip(circleProjectReponame, (username, reponame) -> username + "/" + reponame)
-                .zip(circleBuildNum, (fullPath, buildNum) -> fullPath + "/" + buildNum + "/artifacts/")
-                .zip(circleNodeIndex, (fullPath, nodeIndex) -> fullPath + nodeIndex)
+        Provider<String> externalLocation = getZipper()
+                .zip4(
+                        circleProjectUsername,
+                        circleProjectReponame,
+                        circleBuildNum,
+                        circleNodeIndex,
+                        (username, reponame, buildNum, nodeIndex) ->
+                                username + "/" + reponame + "/" + buildNum + "/artifacts/" + nodeIndex)
                 .zip(circleArtifacts, (fullPath, artifactPath) -> fullPath + artifactPath);
 
-        return physicalArtifactPath.zip(externalLocation, ArtifactLocation::of);
+        Provider<String> circleLink = getZipper()
+                .zip4(
+                        circleUrl,
+                        workflowJobId,
+                        circleNodeIndex,
+                        circleArtifacts,
+                        (url, jobId, nodeIndex, artifactPath) ->
+                                url + "/output/job/" + jobId + "/artifacts/" + nodeIndex + artifactPath);
+
+        return getZipper().zip3(physicalArtifactPath, externalLocation, circleLink, ArtifactLocation::of);
+    }
+
+    private static String extractDomain(String url) {
+        try {
+            URI uri = new URI(url);
+            String scheme = uri.getScheme();
+            String host = uri.getHost();
+
+            if (scheme == null || host == null) {
+                return "Invalid URL";
```

I don't think it ever makes sense to return an invalid state like this as the same type as valid response (ie have a failure just be returned as a string so the caller can't determine it's correct or not). It should throw an exception or return `Optional<String>` and let the caller handle it. I'd throw an exception here - `CIRCLE_BUILD_URL` should _never_ be an invalid url.

---

yeah good point, have made throw instead